django/db/migrations/state.py

```diff
@@ -223,9 +223,38 @@ def __init__(self, real_apps, models, ignore_swappable=False):
         if ignore_swappable:
             pending_models -= {make_model_tuple(settings.AUTH_USER_MODEL)}
         if pending_models:
-            msg = "Unhandled pending operations for models: %s"
-            labels = (".".join(model_key) for model_key in self._pending_operations)
-            raise ValueError(msg % ", ".join(labels))
+            raise ValueError(self._pending_models_error(pending_models))
+
+    def _pending_models_error(self, pending_models):
+        """
+        Almost all internal uses of lazy operations in Django are to resolve
+        string model references in related fields. We can extract the fields
+        from those operations and use them to provide a nicer error message.
+
+        This will work for any function passed to lazy_related_operation that
+        has a keyword argument called 'field'.
+        """
+
+        def extract_field(operation):
+            # Expect a functools.partial with a kwarg called 'field' applied.
+            try:
+                return operation.func.keywords['field']
+            except (AttributeError, KeyError):
+                return None
+
+        extract_fields = lambda ops: list(filter(None, map(extract_field, ops)))
+
+        # Generate pairs of (("app_label", "modelname), [fields])
```

Comment:
`((app_label, model_name), [fields])` LGTM

On April 1, 2015 3:46:12 PM GMT+02:00, Alex Hill <notifications@github.com> wrote:
>> +        from those operations and use them to provide a nicer error
>message.
>> +
>> +        This will work for any function passed to
>lazy_related_operation that
>> +        has a keyword argument called 'field'.
>> +        """
>> +
>> +        def extract_field(operation):
>> +            # Expect a functools.partial with a kwarg called 'field'
>applied.
>> +            try:
>> +                return operation.func.keywords['field']
>> +            except (AttributeError, KeyError):
>> +                return None
>> +
>> +        extract_fields = lambda ops: list(filter(None,
>map(extract_field, ops)))
>> +
>> +        # Generate pairs of (("app_label", "modelname), [fields])
>
>Yes you do. The idea was to give an example of the literal contents of
>the variable rather than its origins, hence the quotes surrounding each
>element. I'm happy to go with ``(app_label, model_name)``, without the
>quotes.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/4423/files#r27569838